pkg/cloudprovider/provider/opennebula/provider.go

```diff
@@ -0,0 +1,430 @@
+	}
+
+	c.Cpu = rawConfig.Cpu
+
+	c.Vcpu = rawConfig.Vcpu
+
+	c.Memory = rawConfig.Memory
+
+	c.Image, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Image)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.Datastore, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Datastore)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.DiskSize = rawConfig.DiskSize
+
+	c.Network, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Network)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.EnableVNC, _, err = p.configVarResolver.GetConfigVarBoolValue(rawConfig.EnableVNC)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.VMTemplateExtra = rawConfig.VMTemplateExtra
+
+	return &c, pconfig, err
+}
+
+func (p *provider) Validate(ctx context.Context, _ *zap.SugaredLogger, spec clusterv1alpha1.MachineSpec) error {
+	_, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %w", err)
+	}
+
+	opennebulaCloudProviderSpec := CloudProviderSpec{}
+	if err = json.Unmarshal(pc.CloudProviderSpec.Raw, &opennebulaCloudProviderSpec); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *provider) GetCloudConfig(_ clusterv1alpha1.MachineSpec) (string, string, error) {
+	return "", "", nil
+}
+
+func (p *provider) Create(ctx context.Context, _ *zap.SugaredLogger, machine *clusterv1alpha1.Machine, data *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c)
+
+	// build a template
+	tpl := vm.NewTemplate()
+
+	// add extra template vars first
+	for key, value := range c.VMTemplateExtra {
+		tpl.Add(keys.Template(key), value)
+	}
+
+	tpl.Add(keys.Name, machine.Spec.Name)
+	tpl.CPU(*c.Cpu).Memory(*c.Memory).VCPU(*c.Vcpu)
+
+	disk := tpl.AddDisk()
+	disk.Add(shared.Image, c.Image)
+	disk.Add(shared.Datastore, c.Datastore)
+	disk.Add(shared.Size, *c.DiskSize)
+
+	nic := tpl.AddNIC()
+	nic.Add(shared.Network, c.Network)
+	nic.Add(shared.Model, "virtio")
+
+	if c.EnableVNC {
+		tpl.AddIOGraphic(keys.GraphicType, "VNC")
+		tpl.AddIOGraphic(keys.Listen, "0.0.0.0")
+	}
+
+	tpl.AddCtx(keys.NetworkCtx, "YES")
+	tpl.AddCtx(keys.SSHPubKey, "$USER[SSH_PUBLIC_KEY]")
+
+	tpl.AddCtx(machineUidContextKey, string(machine.UID))
+	tpl.AddCtx("USER_DATA", base64.StdEncoding.EncodeToString([]byte(userdata)))
+	tpl.AddCtx("USER_DATA_ENCODING", "base64")
+
+	controller := goca.NewController(client)
+
+	// create VM from the generated template above
+	vmID, err := controller.VMs().Create(tpl.String(), false)
+	if err != nil {
+		return nil, err
+	}
+
+	vm, err := controller.VM(vmID).Info(false)
+	if err != nil {
+		return nil, err
```

Comment:
Same here. I'm not _exactly_ sure what this does, but the error should probably look something like this:

```suggestion
		return nil, fmt.Errorf("failed to fetch VM information: %w", err)
```